File: Genetic Algorithm/src/main/java/agents/Vec.java
package agents;// ----------------------------------------------------------------
// The contents of this file are distributed under the CC0 license.
// See http://creativecommons.org/publicdomain/zero/1.0/
// ----------------------------------------------------------------

/// Provides several useful static methods for operating on arrays of doubles
public class Vec
{
	public static Json marshal(double[] vec) {
		Json list = Json.newList();
		for (double v : vec) list.add(v);
		return list;
	}

	public static double[] unmarshal(Json n) {
		double[] vec = new double[n.size()];
		for(int i = 0; i < n.size(); i++)
			vec[i] = n.getDouble(i);
		return vec;
	}

	public static String toString(double[] vec) {
		StringBuilder sb = new StringBuilder();
		if(vec.length > 0) {
			sb.append(Double.toString(vec[0]));
			for(int i = 1; i < vec.length; i++) {
				sb.append(",");
				sb.append(Double.toString(vec[i]));
			}
		}
		return sb.toString();
	}

	public static void setAll(double[] vec, double val) {
		for(int i = 0; i < vec.length; i++)
			vec[i] = val;
	}

	public static double squaredMagnitude(double[] vec) {
		double d = 0.0;
		for (double v : vec) d += v * v;
		return d;
	}

	public static void normalize(double[] vec) {
		double mag = squaredMagnitude(vec);
		if(mag <= 0.0) {
			setAll(vec, 0.0);
			vec[0] = 1.0;
		} else {
			double s = 1.0 / Math.sqrt(mag);
			for(int i = 0; i < vec.length; i++)
				vec[i] *= s;
		}
	}

	public static void copy(double[] dest, double[] src) {
		if(dest.length != src.length)
			throw new IllegalArgumentException("mismatching sizes");
		System.arraycopy(src, 0, dest, 0, src.length);
	}

	public static double[] copy(double[] src) {
		double[] dest = new double[src.length];
		System.arraycopy(src, 0, dest, 0, src.length);
		return dest;
	}

	public static void add(double[] dest, double[] src) {
		if(dest.length != src.length)
			throw new IllegalArgumentException("mismatching sizes");
		for(int i = 0; i < dest.length; i++)
			dest[i] += src[i];
	}

	public static void addScaled(double[] dest, double[] src, double scalar) {
		if(dest.length != src.length)
			throw new IllegalArgumentException("mismatching sizes");
		for(int i = 0; i < dest.length; i++)
			dest[i] += scalar * src[i];
	}

	public static void scale(double[] dest, double scalar) {
		for(int i = 0; i < dest.length; i++)
			dest[i] *= scalar;
	}

	public static double dotProduct(double[] a, double[] b) {
		if(a.length != b.length)
			throw new IllegalArgumentException("mismatching sizes");
		double d = 0.0;
		for(int i = 0; i < a.length; i++)
			d += a[i] * b[i];
		return d;
	}

	public static double squaredDistance(double[] a, double[] b) {
		if(a.length != b.length)
			throw new IllegalArgumentException("mismatching sizes");
		double d = 0.0;
		for(int i = 0; i < a.length; i++) {
			double t = a[i] - b[i];
			d += t * t;
		}
		return d;
	}

	public static void clip(double[] vec, double min, double max) {
		if(max < min)
			throw new IllegalArgumentException("max must be >= min");
		for(int i = 0; i < vec.length; i++)
			vec[i] = Math.max(min, Math.min(max, vec[i]));
	}

	public static double[] concatenate(double[] a, double[] b) {
		double[] c = new double[a.length + b.length];
		System.arraycopy(a, 0, c, 0, a.length);
		System.arraycopy(b, 0, c, a.length, b.length);
		return c;
	}

}

File: Random Forest/src/main/kotlin/helpers/Vec.kt
package helpers

// ----------------------------------------------------------------
// The contents of this file are distributed under the CC0 license.
// See http://creativecommons.org/publicdomain/zero/1.0/
// ----------------------------------------------------------------

import java.lang.Math.max
import java.lang.Math.min

// Provides several useful static methods for operating on arrays of doubles
object Vec {
    fun DoubleArray.marshal(): Json {
        return Json.newList().also {
            for (vec in this)
                it.add(vec)
        }
    }

    fun Json.unmarshal() = DoubleArray(size) { double(it) }

    val DoubleArray.vecString get() = StringBuilder().apply {
        if (isNotEmpty()) {
            append(get(0))
            for (i in 1 until size) {
                append(",").append(get(i))
            }
        }
    }.toString()

    fun DoubleArray.setAll(v: Double) {
        for (i in this.indices)
            this[i] = v
    }

    fun DoubleArray.normalize() {
        val mag = squaredMagnitude
        if (mag <= 0.0) {
            setAll(0.0)
            this[0] = 1.0
        } else {
            val s = 1.0 / Math.sqrt(mag)
            for (i in indices)
                this[i] *= s
        }
    }

    private val DoubleArray.squaredMagnitude get() = sumByDouble { it * it }

    fun DoubleArray.copy(to: DoubleArray) {
        if (to.size != size)
            throw IllegalArgumentException("mismatching sizes")
        mapIndexed { i, _ -> to[i] }
    }

    fun DoubleArray.copy() = DoubleArray(size) { this[it] }

    operator fun DoubleArray.plusAssign(src: DoubleArray) {
        if (size != src.size)
            throw IllegalArgumentException("mismatching sizes")
        mapIndexed { i, it -> it + src[i] }
    }

    operator fun DoubleArray.times(scalar: Double) = DoubleArray(size) { this[it] * scalar }

    operator fun DoubleArray.timesAssign(scalar: Double) {
        map { it * scalar }
    }

    operator fun DoubleArray.times(o: DoubleArray): Double {
        if (size != o.size)
            throw IllegalArgumentException("mismatching sizes")
        return sumByDouble { it * o[indexOf(it)] }
    }

    fun DoubleArray.squaredDistance(o: DoubleArray): Double {
        if (size != o.size)
            throw IllegalArgumentException("mismatching sizes")
        return sumByDouble { it - o[indexOf(it)] }
    }

    fun DoubleArray.clip(min: Double, max: Double) {
        if (max < min)
            throw IllegalArgumentException("max must be >= min")
        map { max(min, min(max, it)) }
    }

    fun DoubleArray.concatenate(o: DoubleArray) = DoubleArray(size + o.size).also {
        System.arraycopy(this, 0, it, 0, size)
        System.arraycopy(o, 0, it, size, o.size)
    }
}
